Add EIP-6120: Universal Token Router (#6120)
ethereum/EIPs: EIPS/eip-draft_universal_router.md

---

> eip: <to be assigned>

Reviewer: ```suggestion
eip: 6120
```

title: Universal Router Contract
description: Universal router contract designed for token allowance that eliminates all `approve` transactions in the future.

> author: Zergity <zergity@gmail.com>

Reviewer: If you'd like to have both your email address and a GitHub username, you can do something like:

```suggestion
author: Zergity (@Zergity), Zergity <zergity@gmail.com>
```

> discussions-to: <URL>

Reviewer: You'll need to open a discussion thread on https://ethereum-magicians.org/

status: Draft
type: Standards Track
category: ERC
created: 2022-12-12

> requires: EIP20, EIP721, EIP1155

Reviewer: ```suggestion
requires: 20, 721, 1155
```

---

## Abstract

> A universal router contract that executes transactions with the following steps:

Reviewer: You should also expand on what a router contract is, what field this is relevant for (EIP-20 tokens, allowances, etc.)

  * (optional) call a calculation contract to get the `amountIn` value, and ensure that this `amountIn` is not greater than an input `amountInMax`
  * transfer `amountIn` of a token from `msg.sender` to a `recipient` address
  * call a contract function to execute an action

>   * (optional) verify the returning amount of a token must be no less than an input `amountOutMin`

Reviewer: This is probably too detailed for the abstract, and might be better in the specification section. Try to give a high level overview, while still remaining technical.

## Motivation

Most Dapp router contracts have the following pattern: approve/permit, (optional) calculation, transferFrom, action, and (optionally) verify the token output. This requires `n*m*k` `allow` (or `permit`) transactions, for `n` Dapps, `m` tokens and `k` user addresses. Even though user approves a contract to spend their tokens, it's the front-end code that they trust, not the contract itself. Anyone can create a front-end code and trick the users to sign a transaction to interact with the Uniswap Router contract and steal all their tokens that have been approved.

Universal Router separates token allowance logic from application logic, acts as a manifest for users when signing a transaction. It saves `(n-1)*m*k` approval transactions for old tokens and **ALL** approval transactions for new tokens. The Universal Router contract is designed to be simple and easy to verify and audit. It's counter-factually `CREATE2`'ed so any new token contracts can hardcode and skip the allowance check entirely.

## Specification

> ```

Reviewer: ````suggestion
```solidity
````

struct Action {
    bool output;    // true for output action, false for input action
    uint eip;       // token type: 0 for ETH, # for ERC#
    address token;  // token contract address
    uint id;        // token id for ERC-721 and ERC-1155
    uint amount;    // amountInMax for input action, amountOutMin for output action
    address recipient;
    address code;   // contract address
    bytes data;     // contract data
}
```

> ```

Reviewer: ````suggestion
```solidity
````

2. `action.eip` specifies the token standard, or `ETH` if it's `0`. If the token is `ETH` and the `recipient` is `0x0`, step #3 is skipped and no transfer is taken, the `amountIn` will be passed to the next output action as the transaction value.
3. Transfer `amountIn` of tokens from `msg.sender` to `action.recipient`.

### Output Action
Actions with `action.output == true` declare the main application action to call, and optionally verify the output token after all is done.
1. If the `action.amount` is not zero, the current token balance of `action.recipient` is recorded for later comparison.
2. Execute the `action.code.call{value: value}(action.data)`, where `value` can be zero or the `amountIn` of the last `ETH` input with an empty `recipient` (see Input Action #2).

### Output Token Verification
After all the actions are handled as above, every token balance tracked in Output Action #2 is queried again for comparison. The balance change must not be less than `action.amount` of each output action, otherwise reverts with "INSUFFICIENT_OUTPUT_AMOUNT".

### Usage Samples
#### UniswapRouter.swapExactTokensForTokens

> ```

Reviewer: ````suggestion
```solidity
````

UniswapV2Helper01.getAmountIn(uint amountOut, address[] memory path) returns (uint amountIn) {
    return UniswapV2Library.getAmountsIn(factory, amountOut, path)[0];
}

// this transaction is signed by user to execute the swap instead of the legacy function
UniversalRouter.exec([{
    output: false,
    eip: 20,
    token: path[0],
    id: 0,
    amount: amountInMax,
    recipient: UniswapV2Library.pairFor(factory, path[0], path[1]),
    code: UniswapV2Helper01.address,
    data: encodeFunctionData("getAmountIn", [amountOut, path]),
}, {
    output: true,
    eip: 20,
    token: path[path.length-1],
    id: 0,
    amount: amountOut,
    recipient: to,
    code: UniswapV2Helper01.address,
    data: encodeFunctionData("swapTokensForExactTokens", [amountOut, amountInMax, path, to, deadline]),
}])
```

## Rationale

The `Permit` type signature is not supported since the purpose of Universal Router is to eliminate all `approve` signatures for new tokens, and *most* for old tokens.

Flashloan transactions are out of scope since it requires support from the application contracts themself.

## Backwards Compatibility

> Old token contracts (ERC20, ERC721 and ERC1155) require approval for Universal Router once for each account.

Reviewer: ```suggestion
Old token contracts (EIP-20, EIP-721 and EIP-1155) require approval for Universal Router once for each account.
```

New token contracts can pre-configure the Universal Router as a trusted spender, and no approval transaction is required.

## Test Cases

Test cases for an implementation are mandatory for EIPs that are affecting consensus changes.  If the test suite is too large to reasonably be included inline, then consider adding it as one or more files in `../assets/eip-####/`.

> 

Reviewer: Remove this section if you don't have test cases.

## Reference Implementation

> ```

Reviewer: ````suggestion
```solidity
````

            return amount;
        }
        if (action.eip == 1155) {
            IERC1155(action.token).safeTransferFrom(msg.sender, action.recipient, action.id, amount, "");
            return amount;
        }
        if (action.eip == 0) {
            if (action.recipient != address(0x0)) {
                TransferHelper.safeTransferETH(action.recipient, amount);
            // } else {
            //     reserved for the next output call value
            }
            return amount;
        }
        revert("UniversalRouter::_input: INVALID_EIP");
    }

    function _balanceOf(address owner, address token, uint eip, uint id) internal view returns (uint balance) {
        if (eip == 20) {
            return IERC20(token).balanceOf(owner);
        }
        if (eip == 1155) {
            return IERC1155(token).balanceOf(owner, id);
        }
        if (eip == 721) {
            return IERC721(token).ownerOf(id) == owner ? 1 : 0;
        }
        if (eip == 0) {
            return owner.balance;
        }
        revert("UniversalRouter::_balanceOf: INVALID_EIP");
    }
}
```

## Security Considerations

> All EIPs must contain a section that discusses the security implications/considerations relevant to the proposed change. Include information that might be important for security discussions, surfaces risks and can be used throughout the life cycle of the proposal. E.g. include security-relevant design decisions, concerns, important discussions, implementation-specific guidance and pitfalls, an outline of threats and risks and how they are being addressed. EIP submissions missing the "Security Considerations" section will be rejected. An EIP cannot proceed to status "Final" without a Security Considerations discussion deemed sufficient by the reviewers.

Reviewer: ```suggestion
TODO
```
